Sources/Crust/Shine.swift

```diff
@@ -0,0 +1,113 @@
+private enum LayoutBlock {
+  case Implicit
+  case Explicit
+
+  var isImplicit : Bool {
+    switch self {
+    case .Implicit: return true
+    default: return false
+    }
+  }
+}
+
+/// Process a raw Silt token stream into a Stainless Silt token stream by
+/// inserting layout markers in the appropriate places.  This ensures that we
+/// have an explicitly-scoped input to the Parser before we even try to do a
+/// Scope Check.
+public func layout(_ ts : [TokenSyntax]) -> [TokenSyntax] {
+  var toks = ts
+  var stainlessToks = [TokenSyntax]()
+  var layoutBlockStack = [LayoutBlock]()
+  while !toks.isEmpty && toks[0].tokenKind != .eof {
+    // Pop the first token in the stream
+    let tok = toks.removeFirst()
+
+    // If we've got a layout word, check to see if we need to push a new
+    // layout block
+    if tok.tokenKind == .whereKeyword || tok.tokenKind == .fieldKeyword {
+      stainlessToks.append(tok)
+
+      guard let peekTok = toks.first else {
+        // If there's nothing after the layout word, open an implicit block and
+        // append a brace into the token stream.
+        layoutBlockStack.append(.Implicit)
+        stainlessToks.append(TokenSyntax(.leftBrace, presence: .implicit))
+        continue
+      }
+
+      if peekTok.tokenKind == .leftBrace {
+        layoutBlockStack.append(.Explicit)
+      } else {
+        layoutBlockStack.append(.Implicit)
+        stainlessToks.append(TokenSyntax(.leftBrace, presence: .implicit))
+      }
+      continue
+    }
+
+    // Next, check to see if we have a closing brace to match anything on
+    // the layout stack
+    if tok.tokenKind == .rightBrace {
+      // If we're inside an implicit layout block but see an explicit closer,
+      // we need to pop off as many implicit layout blocks as it takes to
+      // find the matching explicit layout block
+      assert(!layoutBlockStack.isEmpty,
+             """
+             Empty layout stack encountered while trying to match explicit '}' \
+             at \(tok.sourceRange as Optional)
+             """)
+      var foundExplicit = false
+      while let implTop = layoutBlockStack.popLast() {
+        switch implTop {
+        case .Implicit:
+          stainlessToks.append(TokenSyntax(.rightBrace, presence: .implicit))
+        case .Explicit:
+          foundExplicit = true
+          break
+        }
+      }
+      assert(foundExplicit,
+             """
+             Empty layout stack encountered while trying to match explicit '}' \
+             at \(tok.sourceRange as Optional)
+             """)
+    }
+
+    stainlessToks.append(tok)
+
+    // Finally, check to see if we've got a token on a new line.  If so, we
+    // need to insert a semicolon in the token stream if we can't find one.
+    //
+    // FIXME: This seems wrong in general.
+    if
+      let peekTok = toks.first,
+      let tokRange = tok.sourceRange,
+      let peekRange = peekTok.sourceRange,
+      tokRange.start.line != peekRange.start.line,
```

Comment:
Check the Trivia instead?

Reply:
I would have to search the trivia on each token for a newline, right?